lib/kazan/watcher.ex

```diff
@@ -0,0 +1,164 @@
+          process_job_event(type, job)
+      end
+      {:noreply, state}
+    end
+  ```
+  """
+
+  use GenServer
+  alias Kazan.LineBuffer
+  require Logger
+
+  defmodule State do
+    @moduledoc """
+    Stores the internal state of the watcher.
+
+    Includes:
+
+    resource_version:
+
+    The K8S *resource_version* (RV) is used to version each change in the cluster.
+    When we listen for events we need to specify the RV to start listening from.
+    For each event received we also receive a new RV, which we store in state.
+    When the watch eventually times-out (it is an HTTP request with a chunked response),
+    we can then create a new request passing the latest RV that was received.
+
+    buffer:
+
+    Chunked HTTP responses are not always a complete line, so we must buffer them
+    until we have a complete line before parsing.
+    """
+    defstruct id: nil, request: nil, send_to: nil, buffer: LineBuffer.new(), rv: nil
+  end
+
+  @doc "Starts an async request send_to is the process to send messages to; resource_version can also be passed"
+  def start_link(%Kazan.Request{} = request, opts) do
+    send_to = Keyword.fetch!(opts, :send_to)
+    rv = Keyword.fetch!(opts, :resource_version)
+    GenServer.start_link(__MODULE__, [request, rv, send_to])
+  end
+
+  @impl GenServer
+  def init([request, rv, send_to]) do
+    Logger.info "Watcher init pid: #{inspect self()} rv: #{rv} request: #{inspect request}"
+    {:ok, id} = Kazan.Client.run(request, stream_to: self())
+    {:ok, %State{id: id, request: request, rv: rv, send_to: send_to}}
+  end
+
+  @impl GenServer
+  def handle_info(%HTTPoison.AsyncStatus{code: 200}, state) do
+    {:noreply, state}
+  end
+
+  @impl GenServer
+  def handle_info(%HTTPoison.AsyncHeaders{}, state) do
+    {:noreply, state}
+  end
+
+  @impl GenServer
+  def handle_info(%HTTPoison.AsyncChunk{chunk: chunk, id: request_id},
+                  %State{id: id, buffer: buffer, rv: rv, send_to: send_to} = state) do
+    if request_id != id do
+      Logger.warn "Received chunk for wrong request_id: #{inspect request_id} expect: #{inspect id}"
+      {:noreply, state}
+    else
+
+      {lines, buffer} =
+        buffer
+        |> LineBuffer.add_chunk(chunk)
+        |> LineBuffer.get_lines()
+
+        new_rv = process_lines(lines, rv, send_to)
+
+      {:noreply, %State{state | buffer: buffer, rv: new_rv}}
+    end
+  end
+
+  @impl GenServer
+  def handle_info(%HTTPoison.Error{reason: {:closed, :timeout}}, state) do
+    Logger.info "Received Timeout with id: #{inspect self()}"
+    {:noreply, restart_connection(state)}
+  end
+
+  @impl GenServer
+  def handle_info(%HTTPoison.AsyncEnd{id: request_id},
+                  %State{id: id} = state) do
+    Logger.info "Received AsyncEnd for id: #{inspect request_id}"
+    if request_id != id do
+      Logger.warn "Received AsyncEnd message for wrong request_id: #{inspect request_id} expect: #{inspect id}"
+      {:noreply, state}
```

Comment:
Again, this shouldn't really happen, so I'm wondering if the correct thing to do would be to let it crash so supervisor can restart things in a hopefully-good state.

Reply:
Yes